pkg/trace/api/normalizer.go

```diff
@@ -21,48 +23,81 @@ const (
 	MaxTypeLen = 100
-	// MaxEndDateOffset the maximum amount of time in the future we
-	// tolerate for span end dates
-	MaxEndDateOffset = 10 * time.Minute
+	// DefaultServiceName TODO: do we really want such indescriptive defaults here like "service"?
+	DefaultServiceName = "no_name_service"
+	// DefaultSpanName TODO: do we really want such indescriptive defaults here like "service.trace"?
+	DefaultSpanName = "no_name_span"
 )
 
 var (
 	// Year2000NanosecTS is an arbitrary cutoff to spot weird-looking values
 	Year2000NanosecTS = time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC).UnixNano()
 )
 
-// normalize makes sure a Span is properly initialized and encloses the minimum required info
-func normalize(s *pb.Span) error {
+// normalize makes sure a Span is properly initialized and encloses the minimum required info, returning error if it
+// is invalid beyond repair
+func normalize(ts *info.TagStats, s *pb.Span) error {
+	fallbackServiceName := DefaultServiceName
+	if ts.Lang != "" {
+		fallbackServiceName = ts.Lang
+	}
+
+	// While it's technically possible to get TraceID=0 (with 1 in 2**64 probability) we figured it's far more likely
+	// for an integration coding error to produce TraceID=0 so we prevent that here to avoid negative downstream
+	// consequences. Same thing for SpanID.
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		return fmt.Errorf("dropping invalid trace (reason:trace_id_zero): %s", s)
+	}
+	if s.SpanID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)
+		return fmt.Errorf("dropping invalid trace (reason:span_id_zero): %s", s)
+	}
+
 	// Service
 	if s.Service == "" {
-		return errors.New("empty `Service`")
+		atomic.AddInt64(&ts.TracesMalformed.ServiceEmpty, 1)
+		log.Debugf("Fixing malformed trace (reason:service_empty), setting span.service=%s: %s", fallbackServiceName, s)
+		s.Service = fallbackServiceName
 	}
 	if len(s.Service) > MaxServiceLen {
-		return fmt.Errorf("`Service` too long (%d chars max): %s", MaxServiceLen, s.Service)
+		atomic.AddInt64(&ts.TracesMalformed.ServiceTruncate, 1)
+		log.Debugf("Fixing malformed trace (reason:service_truncate), truncating span.service to length=%d: %s", MaxServiceLen, s)
+		s.Service = s.Service[:MaxServiceLen]
 	}
 	// service should comply with Datadog tag normalization as it's eventually a tag
 	svc := normalizeTag(s.Service)
 	if svc == "" {
-		return fmt.Errorf("invalid `Service`: %q", s.Service)
+		atomic.AddInt64(&ts.TracesMalformed.ServiceInvalid, 1)
+		log.Debugf("Fixing malformed trace (reason:service_invalid), replacing invalid span.service=%s with fallback span.service=%s: %s", s.Service, fallbackServiceName, s)
+		svc = fallbackServiceName
 	}
 	s.Service = svc
 
 	// Name
 	if s.Name == "" {
-		return errors.New("empty `Name`")
+		atomic.AddInt64(&ts.TracesMalformed.SpanNameEmpty, 1)
+		log.Debugf("Fixing malformed trace (reason:span_name_empty), setting span.name=%s: %s", DefaultSpanName, s)
```

Comment:
Looks like it does, here's how it gets formatted as `%s`: `Fixing malformed trace (reason:service_empty), setting span.service=service: name:"django.controller" resource:"GET /some/raclette" traceID:424242 spanID:2382721542312139272 parentID:1111 start:1448466874000000000 duration:10000000 meta:<key:"pool" value:"fondue" > meta:<key:"user" value:"leo" > metrics:<key:"cheese_weight" value:100000 > type:"http"`


And `%#v`: `Fixing malformed trace (reason:service_empty), setting span.service=service: &pb.Span{Service:"", Name:"django.controller", Resource:"GET /some/raclette", TraceID:0x67932, SpanID:0xb4e82abb8e515c1f, ParentID:0x457, Start:1448466874000000000, Duration:10000000, Error:0, Meta:map[string]string{"pool":"fondue", "user":"leo"}, Metrics:map[string]float64{"cheese_weight":100000}, Type:"http"}`

Changed to `%#v` though as that is a bit clearer since the whole span is wrapped in `&pb.Span{...}`.

Reply:
Actually no, let's leave it with `%s` as that's perfect. It was designed to be user friendly. `#%v` will only make sense to Go users, while `%s` is more language agnostic.

Reply:
Reverted